gobblin-runtime/src/main/java/org/apache/gobblin/runtime/TaskStateCollectorService.java

```diff
@@ -206,12 +222,53 @@ public Void call() throws Exception {
 
     LOGGER.info(String.format("Collected task state of %d completed tasks", taskStateQueue.size()));
 
+    Long taskByteSize;
+    String stringSize;
+    Double newPercentageCopied;
     // Add the TaskStates of completed tasks to the JobState so when the control
     // returns to the launcher, it sees the TaskStates of all completed tasks.
     for (TaskState taskState : taskStateQueue) {
       consumeTaskIssues(taskState);
       taskState.setJobState(this.jobState);
       this.jobState.addTaskState(taskState);
+
+      if (jobState.getPropAsBoolean(ConfigurationKeys.REPORT_JOB_COPY_PROGRESS, false)) {
+        stringSize = taskState.getProp(ServiceConfigKeys.WORK_UNIT_BYTE_SIZE);
+        if (stringSize == null) {
+          LOGGER.warn("Expected to report data copy progress but work unit byte size property null");
+          break;
+        }
+
+        taskByteSize = Long.parseLong(stringSize);
+        // if progress reporting is enabled, value should be present
+        String strTotalSizeToCopy = this.jobState.getProp(AbstractJobLauncher.TOTAL_BYTES_TO_COPY);
+        if (strTotalSizeToCopy == null) {
+          LOGGER.warn("Expected to report data copy progress but total bytes to copy property null");
+          break;
+        }
+
+        this.totalSizeToCopy = Long.parseLong(strTotalSizeToCopy);
+        // avoid flooding Kafka message queue by sending GobblinTrackingEvents only when threshold is passed
+        this.bytesCopiedSoFar += taskByteSize;
+        newPercentageCopied = this.bytesCopiedSoFar / this.totalSizeToCopy;
+        if (newPercentageCopied - this.lastPercentageReported > ConfigurationKeys.DEFAULT_PROGRESS_REPORTING_THRESHOLD) {
+          this.lastPercentageReported = newPercentageCopied;
+          Long percentageToReport = (long) (this.lastPercentageReported * 100);
```

Comment:
Should we do a Math.Round() here? I think by default the value will be truncated when you convert double to long. so 99.999 will be 99 , not 100. And we can end up with jobs that are fully complete, but still show 99%  progress.